File: verify/iscostas.js

// TODO: documentation

const text = await new Response(Deno.stdin.readable).text();
const lines = text.trim().split("\n");
const n = parseInt(lines[0]);
const perm = lines[1].trim().split(/\s+/).map(Number);

function iscostas(perm) {
    const n = perm.length;
    if (new Set(perm).size !== n) return false;
    for (let k = 1; k < n; k++) {
        const seen = new Set();
        for (let i = 0; i < n - k; i++) {
            const d = perm[i + k] - perm[i];
            if (seen.has(d)) return false;
            seen.add(d);
        }
    }
    return true;
}

console.log(iscostas(perm) ? 1 : 0);

File: verify/iscostas.clj
;; TODO: documentation

(defn iscostas [perm]
  (let [n (count perm)
        v (vec perm)]
    (and (= (count (set perm)) n)
         (every? (fn [k]
                   (let [diffs (map (fn [i] (- (v (+ i k)) (v i)))
                                    (range (- n k)))]
                     (= (count (set diffs)) (count diffs))))
                 (range 1 n)))))



(let [_n   (Integer/parseInt (clojure.string/trim (read-line)))
      perm (map #(Integer/parseInt %)
                (clojure.string/split (clojure.string/trim (read-line)) #"\s+"))]
  (println (if (iscostas perm) 1 0)))
